File: Leetcode/dynamic_programming/53_max_subarray/Solution.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if max(nums) < 0: return max(nums)
        maxSoFar, maxEndingHere = 0, 0
        for num in nums:
            maxEndingHere = max(0, maxEndingHere+num)
            maxSoFar = max(maxSoFar, maxEndingHere)
        return maxSoFar

File: Leetcode/dynamic_programming/53_max_subarray/Solution.java
class Solution {
    public int maxSubArray(int[] A) {
        int maxSoFar=Integer.MIN_VALUE, maxEndingHere=0;
        for (int i=0;i<A.length;++i){
            maxEndingHere= Math.max(maxEndingHere+A[i],A[i]);
            maxSoFar=Math.max(maxSoFar, maxEndingHere);
        }
        return maxSoFar;
    }
}
